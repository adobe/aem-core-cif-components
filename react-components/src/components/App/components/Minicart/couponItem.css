.root {
    display: grid;
    grid-gap: 0 1rem;
    grid-template-columns: 1fr min-content;
    margin: 0 1.5rem;
}
.couponName {
    line-height: 1.5rem;
}
