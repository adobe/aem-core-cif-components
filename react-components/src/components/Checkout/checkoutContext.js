/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React, { createContext, useContext, useReducer } from 'react';
import { string, func, shape, object } from 'prop-types';

export const initialCheckoutState = {
    flowState: 'cart',
    order: null,
    editing: null,
    shippingAddress: null,
    billingAddress: null,
    shippingMethod: null,
    paymentMethod: null,
    braintreeToken: false
};

export const checkoutReducer = (state, action) => {
    switch (action.type) {
        case 'beginCheckout':
            return {
                ...state,
                flowState: 'form'
            };
        case 'cancelCheckout':
            return {
                ...state,
                flowState: 'cart'
            };
        case 'placeOrder':
            return {
                ...state,
                order: action.order,
                flowState: 'receipt',
                editing: 'receipt'
            };
        case 'reset':
            return {
                ...initialCheckoutState
            };
        case 'setEditing':
            return {
                ...state,
                editing: action.editing
            };
        case 'endEditing':
            return {
                ...state,
                editing: null
            };
        case 'setShippingAddress':
            return {
                ...state,
                shippingAddress: action.shippingAddress,
                editing: null
            };
        case 'setBillingAddress':
            return {
                ...state,
                billingAddress: action.billingAddress,
                editing: null
            };
        case 'setShippingMethod':
            return {
                ...state,
                shippingMethod: action.shippingMethod,
                editing: null
            };
        case 'setPaymentMethod':
            return {
                ...state,
                paymentMethod: action.paymentMethod,
                editing: null
            };
        case 'setBraintreeToken':
            return {
                ...state,
                braintreeToken: action.token
            };

        default:
            return state;
    }
};

export const CheckoutContext = createContext();

export const CheckoutProvider = props => {
    const reducer = props.reducer || checkoutReducer;
    const initialState = props.initialState || initialCheckoutState;
    return (
        <CheckoutContext.Provider value={useReducer(reducer, initialState)}>{props.children}</CheckoutContext.Provider>
    );
};

CheckoutProvider.propTypes = {
    reducer: func,
    initialState: shape({
        flowState: string.isRequired,
        order: object,
        editing: object,
        shippingAddress: object,
        billingAddress: object,
        shippingMethod: object,
        paymentMethod: object
    })
};

export const useCheckoutState = () => useContext(CheckoutContext);
