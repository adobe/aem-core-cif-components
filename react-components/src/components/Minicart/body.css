.root {
    display: grid;
    grid-template-rows: min-content min-content 1fr;
    overflow: auto;
}
