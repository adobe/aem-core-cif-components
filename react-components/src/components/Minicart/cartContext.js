/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React, { createContext, useContext, useReducer } from 'react';
import { object, func } from 'prop-types';
import { useCookieValue } from '../../utils/hooks';

export const initialState = {
    isOpen: false,
    isRegistered: false,
    isEditing: false,
    isLoading: false,
    editItem: {},
    cartId: null,
    cart: null,
    errorMessage: null,
    couponError: null
};

export const reducerFactory = setCartCookie => {
    return (state, action) => {
        switch (action.type) {
            case 'close':
                return {
                    ...state,
                    isOpen: false
                };
            case 'open':
                return {
                    ...state,
                    isOpen: true
                };
            case 'beginLoading':
                return {
                    ...state,
                    isLoading: true
                };
            case 'endLoading':
                return {
                    ...state,
                    isLoading: false
                };
            case 'beginEditing':
                return {
                    ...state,
                    isEditing: true,
                    editItem: action.item
                };
            case 'endEditing':
                return {
                    ...state,
                    isEditing: false,
                    editItem: {}
                };
            case 'cartId':
                return {
                    ...state,
                    cartId: action.cartId
                };
            case 'reset':
                setCartCookie('', 0);
                return {
                    ...state,
                    cartId: null,
                    errorMessage: null,
                    couponError: null,
                    cart: null
                };
            case 'cart':
                return {
                    ...state,
                    cart: action.cart,
                    isLoading: false,
                    couponError: !action.cart.applied_coupon ? state.couponError : null
                };
            case 'register': {
                return {
                    ...state,
                    isRegistered: true
                };
            }
            case 'error':
                console.error(action.error);
                return {
                    ...state,
                    errorMessage: action.error
                };
            case 'discardError':
                return {
                    ...state,
                    errorMessage: null
                };
            case 'couponError':
                return {
                    ...state,
                    couponError: action.error
                };

            default:
                return state;
        }
    };
};

export const CartContext = createContext();

export const CartProvider = props => {
    const factory = props.reducerFactory || reducerFactory;
    const state = props.initialState || initialState;

    const [, setCartCookie] = useCookieValue('cif.cart');
    const contextValue = useReducer(factory(setCartCookie), state);

    return <CartContext.Provider value={contextValue}>{props.children}</CartContext.Provider>;
};

CartProvider.propTypes = {
    reducerFactory: func,
    initialState: object
};

export const useCartState = () => useContext(CartContext);
