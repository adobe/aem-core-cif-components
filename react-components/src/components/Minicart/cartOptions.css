/*******************************************************************************
 *
 *    Copyright 2021 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
.root {
    display: grid;
    grid-template-rows: min-content 1fr;
    height: calc(100% - var(--minicart-header-height));
    left: 0;
    position: absolute;
    top: 3.5rem;
    width: 100%;
}

.focusItem {
    box-shadow: 0 1px rgb(var(--venia-border));
    display: grid;
    gap: 2rem;
    grid-auto-flow: column;
    grid-template-columns: 1fr;
    padding: 1rem;
}

.name {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.form {
    overflow: auto;
}

.modal {
    visibility: hidden;
    height: 100%;
    width: 100%;
    background-color: rgb(var(--venia-grey));
    text-align: center;
    position: absolute;
    bottom: 0;
}

.modal_active {
    visibility: visible;
    composes: modal;
    opacity: 0.9;
}

.options {
    font-weight: bold;
}

.quantity {
    border-color: rgb(var(--venia-border));
    border-style: solid;
    border-width: 0 0 1px;
    margin: 0 1.5rem;
    padding-top: 1rem;
    padding-bottom: 1rem;
}

.quantityTitle {
    font-size: 1rem;
    font-weight: 600;
    line-height: 1.5;
    margin-bottom: 1.5rem;
}

.save {
    align-items: center;
    background-color: white;
    display: grid;
    grid-auto-columns: min-content;
    grid-auto-flow: column;
    grid-gap: 0.75rem;
    height: 5rem;
    justify-content: center;
    justify-items: center;
    position: relative;
}
