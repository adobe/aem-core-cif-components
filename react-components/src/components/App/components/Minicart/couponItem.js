/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { useCartState } from './cartContext';
import Kebab from './kebab';
import Section from './section';

import classes from './couponItem.css';

const CouponItem = () => {
    const [{ cart, removeCoupon }] = useCartState();

    const appliedCoupon = cart.applied_coupon ? cart.applied_coupon.code : null;

    return (
        <div className={classes.root}>
            <div className={classes.couponName}>
                Coupon <strong>{appliedCoupon}</strong> applied.
            </div>
            <Kebab>
                <Section text="Remove coupon" onClick={removeCoupon} icon="Trash" />
            </Kebab>
        </div>
    );
};

export default CouponItem;
