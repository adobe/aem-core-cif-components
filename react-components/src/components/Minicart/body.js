/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { useTranslation } from 'react-i18next';

import LoadingIndicator from '../LoadingIndicator';

import EmptyMinicartBody from './emptyMinicartBody';
import classes from './body.css';
import ProductList from './productList';
import CartOptions from './cartOptions';
import Error from './error';
import CouponForm from './couponForm';
import CouponItem from './couponItem';

import { useCartState } from './cartContext';
import DiscountList from './discountList';

const CartLoadingIndicator = () => {
    const [t] = useTranslation('cart');

    return <LoadingIndicator>{t('cart:fetching-data', 'Fetching cart data...')}</LoadingIndicator>;
};

const Body = () => {
    const [{ isEditing, cart, isLoading, errorMessage }] = useCartState();
    const isEmpty = cart && Object.entries(cart).length > 0 ? cart.items.length === 0 : true;

    if (isLoading) {
        return <CartLoadingIndicator />;
    }

    if (errorMessage) {
        return <Error />;
    }

    if (isEmpty) {
        return <EmptyMinicartBody />;
    }

    if (isEditing) {
        return <CartOptions />;
    }

    const cartItems = cart.items;
    const couponFragment = !cart.applied_coupon ? <CouponForm /> : <CouponItem />;
    const discountFragment =
        cart.prices.discounts && cart.prices.discounts.length > 0 ? (
            <DiscountList discounts={cart.prices.discounts} />
        ) : (
            <div />
        );

    return (
        <div className={classes.root}>
            <ProductList cartItems={cartItems} />
            {discountFragment}
            {couponFragment}
        </div>
    );
};

export default Body;
