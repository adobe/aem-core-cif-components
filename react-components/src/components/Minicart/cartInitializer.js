/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useEffect } from 'react';
import { useCartState } from './cartContext';
import { useCookieValue } from '../../utils/hooks';
import { useUserContext } from '../../context/UserContext';

const CartInitializer = props => {
    const [{ cartId: stateCartId }, dispatch] = useCartState();
    const [{ cartId: registeredCartId }] = useUserContext();

    const CART_COOKIE = 'cif.cart';

    const [cartId, setCartCookie] = useCookieValue(CART_COOKIE);

    useEffect(() => {
        if (cartId && cartId.length > 0 && !stateCartId) {
            dispatch({ type: 'cartId', cartId });
        }
    }, [cartId]);

    useEffect(() => {
        if (stateCartId && (!cartId || cartId.length === 0)) {
            setCartCookie(stateCartId);
        }
    }, [stateCartId]);

    useEffect(() => {
        if (registeredCartId) {
            setCartCookie(registeredCartId);
            dispatch({
                type: 'cartId',
                cartId: registeredCartId
            });
        }
    }, [registeredCartId]);

    return props.children;
};

export default CartInitializer;
