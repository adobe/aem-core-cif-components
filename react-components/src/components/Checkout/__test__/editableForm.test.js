/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { wait } from '@testing-library/react';
import { render } from '../../../utils/test-utils';
import EditableForm from '../editableForm';
import { CartProvider } from '../../Minicart';
import { CheckoutProvider } from '../checkoutContext';

import CREATE_BRAINTREE_CLIENT_TOKEN from '../../../queries/mutation_create_braintree_client_token.graphql';
import QUERY_COUNTRIES from '../../../queries/query_countries.graphql';

describe('<EditableForm />', () => {
    const mocksQueryCountries = [
        {
            request: {
                query: QUERY_COUNTRIES
            },
            result: {
                data: {
                    countries: [
                        {
                            id: 'US',
                            available_regions: [
                                { id: 4, code: 'AL', name: 'Alabama' },
                                { id: 7, code: 'AK', name: 'Alaska' }
                            ]
                        }
                    ]
                }
            }
        }
    ];
    it('renders the shipping address form if countries are loaded', async () => {
        const { queryByText } = render(
            <CartProvider initialState={{}} reducerFactory={() => state => state}>
                <CheckoutProvider initialState={{ editing: 'address', flowState: 'form' }} reducer={state => state}>
                    <EditableForm />
                </CheckoutProvider>
            </CartProvider>,
            { mocks: mocksQueryCountries }
        );

        await wait(() => {
            expect(queryByText('Shipping Address')).not.toBeNull();
        });
    });

    it('renders the payments form if countries are loaded', async () => {
        const mocksPaymentsForm = [
            ...mocksQueryCountries,
            {
                request: {
                    query: CREATE_BRAINTREE_CLIENT_TOKEN
                },
                result: {
                    data: {
                        createBraintreeClientToken: 'my-sample-token'
                    }
                }
            }
        ];

        const mockCartState = {
            cart: {
                available_payment_methods: [
                    {
                        code: 'braintree',
                        title: 'Credit Card (Braintree)'
                    },
                    {
                        code: 'checkmo',
                        title: 'Check / Money order'
                    }
                ],
                is_virtual: false
            }
        };

        let mockReducer = jest.fn(state => state);
        const { queryByText } = render(
            <CartProvider initialState={mockCartState} reducerFactory={() => state => state}>
                <CheckoutProvider initialState={{ editing: 'paymentMethod', flowState: 'form' }} reducer={mockReducer}>
                    <EditableForm />
                </CheckoutProvider>
            </CartProvider>,
            { mocks: mocksPaymentsForm }
        );

        await wait(() => {
            expect(queryByText('Billing Information')).not.toBeNull();
            expect(mockReducer.mock.calls.length).toBe(1);
        });
    });

    it('renders the shipping method form if countries are loaded', async () => {
        const { queryByText } = render(
            <CartProvider initialState={{}} reducerFactory={() => state => state}>
                <CheckoutProvider
                    initialState={{ editing: 'shippingMethod', flowState: 'form' }}
                    reducer={state => state}>
                    <EditableForm />
                </CheckoutProvider>
            </CartProvider>,
            { mocks: mocksQueryCountries }
        );

        await wait(() => {
            expect(queryByText('Shipping Information')).not.toBeNull();
        });
    });

    it('does not render the shipping address form if countries could not be loaded', async () => {
        const mocks = [
            {
                request: {
                    query: QUERY_COUNTRIES
                },
                result: {
                    data: {
                        countries: []
                    }
                }
            }
        ];

        const { asFragment } = render(
            <CartProvider initialState={{}} reducerFactory={() => state => state}>
                <CheckoutProvider initialState={{ editing: 'address', flowState: 'form' }} reducer={state => state}>
                    <EditableForm />
                </CheckoutProvider>
            </CartProvider>,
            { mocks: mocks }
        );

        await wait(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
