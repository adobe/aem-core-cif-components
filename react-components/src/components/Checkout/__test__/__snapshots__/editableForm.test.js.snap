// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableForm /> does not render the shipping address form if countries could not be loaded 1`] = `<DocumentFragment />`;
